### PR TITLE
Add show_from profile filter.

### LINKED ISSUE #363: Add support for show-from filter
When viewing at large profiles in a tree form (e.g. flame graph or a top-down tree), it is sometimes desired to "re-root" the profile on a given function of interest. E.g. if as a developer I am investigating the slowness of foo() function, I'd like to look at aggregated data where all callers of foo() are ignored and the data is a tree starting from foo() and expanding to its callees.

To support this, add support for `--show-from` flag which will take a regular expression (similar to current show and hide filters) and will remove all the callers up to (but excluding) the first matching frame in every stack (in the caller -> callee order).

For example:
```
main()
  -> bar1()
    -> foo()
      -> impl1()
      -> impl2()
        -> impl21()
          -> malloc()
    -> bar12()
      -> foo()
        -> impl2()
  -> bar2()
    -> foo()
      -> impl1()
```
Applying `--show-from=foo` will give:
```
foo()
  -> impl1()
  -> impl2()
    -> impl21()
      -> malloc()
```

The removal is done up to last matching from, so specifying `--show-from=foo|impl` would still give the same result rather than
```
impl1()
impl21()
  -> malloc()
```
This is so that specifying `--show` filter for the same regular expression does not result in more matching frames being visible than `--show-from`.

Like `show` and `from` filters, the regular expression should match function name, source file name and mapping path. Locations may be matched partially in case of inline lines presence.

@rauls5382 FYI

### PR DESCRIPTION
**Pull request by lvng**
This change adds the show_from filter, which drops frames before the first match in each stack, beginning from the root.

This fixes #363.

### CODE REVIEW

File: internal/driver/commands.go [hunk 1]
```diff
@@ -185,6 +185,10 @@ var pprofVariables = variables{
 		"Only show nodes matching regexp",
 		"If set, only show nodes that match this location.",
 		"Matching includes the function name, filename or object name.")},
+	"show_from": &variable{stringKind, "", "", helpText(
+		"Drops nodes above the highest matched node of every sample.",
```
[the remark this answers is not in the document]
Comment (lvng) [reply internal/driver/commands.go]: done.

File: internal/driver/driver.go [hunk 2]
```diff
@@ -193,6 +195,9 @@ func applyCommandOverrides(cmd []string, v variables) variables {
 		v.set("hide", "")
 		v.set("show", "")
 	}
+	if !showFrom {
+		v.set("show_from", "")
```
[the remark this answers is not in the document]
Comment (lvng) [reply internal/driver/driver.go]: I wasn't sure whether to include it in the focus or hide section. it seems those flags are semantically equivalent so i combined focus, hide, and showFrom into a single flag "filter".

File: internal/driver/driver_test.go [hunk 3]
```diff
@@ -247,6 +248,7 @@ func testSourceURL(port int) string {
 
 // solutionFilename returns the name of the solution file for the test
 func solutionFilename(source string, f *testFlags) string {
+	fmt.Printf("solution flags: %v\n", f)
```
[the remark this answers is not in the document]
Comment (lvng) [reply internal/driver/driver_test.go]: yup. fixed.

File: profile/filter.go [hunk 4]
```diff
@@ -74,6 +74,97 @@ func (p *Profile) FilterSamplesByName(focus, ignore, hide, show *regexp.Regexp)
 	return
 }
 
+// ShowFrom drops all stack frames before the first match from the root and
```
[the remark this answers is not in the document]
Comment (lvng) [reply profile/filter.go]: changed to "highest"

File: profile/filter.go [hunk 5]
```diff
@@ -74,6 +74,97 @@ func (p *Profile) FilterSamplesByName(focus, ignore, hide, show *regexp.Regexp)
...
+// returns whether a match was found. If showFrom is nil it returns false and
+// does not modify the profile.
+//
+// Example: consider a sample with frames [A, B, C], where A is the root.
```
[the remark this answers is not in the document]
Comment (lvng) [reply profile/filter.go]: Added a duplicate B frame to illustrate matching multiple frames.

File: profile/filter.go [hunk 6]
```diff
@@ -74,6 +74,97 @@ func (p *Profile) FilterSamplesByName(focus, ignore, hide, show *regexp.Regexp)
...
+// ShowFrom(nil) returns false and has frames [A, B, C].
+// ShowFrom(A) returns true and has frames [A, B, C].
+// ShowFrom(B) returns true and has frames [B, C].
+// ShowFrom(C) returns true and has frames [C].
+// ShowFrom(D) returns false and the sample is dropped because no frames remain.
+func (p *Profile) ShowFrom(showFrom *regexp.Regexp) (matched bool) {
+	if showFrom == nil {
+		return false
+	}
+
```
[the remark this answers is not in the document]
Comment (lvng) [reply profile/filter.go]: done.

File: profile/filter.go [hunk 7]
```diff
@@ -74,6 +74,97 @@ func (p *Profile) FilterSamplesByName(focus, ignore, hide, show *regexp.Regexp)
...
+	p.Sample = filteredSamples
+	return
+}
+
+// filterShowFromLocation tests a showFrom regex against a location and removes
+// lines after the last match. If the mapping is matched, then all lines are
+// kept.
+func filterShowFromLocation(location *Location, showFrom *regexp.Regexp) (matched bool) {
+	if showFrom == nil {
```
[the remark this answers is not in the document]
Comment (lvng) [reply profile/filter.go]: done.

File: profile/filter.go [hunk 8]
```diff
@@ -74,6 +74,97 @@ func (p *Profile) FilterSamplesByName(focus, ignore, hide, show *regexp.Regexp)
...
+	filteredSamples := make([]*Sample, 0, len(p.Sample))
+	for _, sample := range p.Sample {
+		if i := lastLocationIndex(sample.Location, showFromLocs); i >= 0 {
+			sample.Location = sample.Location[0 : i+1]
+			filteredSamples = append(filteredSamples, sample)
+		}
+	}
+
+	p.Sample = filteredSamples
+	return
```
[the remark this answers is not in the document]
Comment (lvng) [reply profile/filter.go]: done.

File: profile/filter.go [hunk 9]
```diff
@@ -74,6 +74,97 @@ func (p *Profile) FilterSamplesByName(focus, ignore, hide, show *regexp.Regexp)
...
+				return i
+			}
+		}
+	}
+	return -1
+}
+
+// matchesMapping returns whether a regex matches a location's mapping.
+func (loc *Location) matchesMapping(re *regexp.Regexp) bool {
+	if m := loc.Mapping; m != nil && re != nil && re.MatchString(m.File) {
```
[the remark this answers is not in the document]
Comment (lvng) [reply profile/filter.go]: done.

File: profile/filter.go [hunk 10]
```diff
@@ -74,6 +74,97 @@ func (p *Profile) FilterSamplesByName(focus, ignore, hide, show *regexp.Regexp)
...
+func (p *Profile) ShowFrom(showFrom *regexp.Regexp) (matched bool) {
+	if showFrom == nil {
+		return false
+	}
+
+	// showFromLocs stores location IDs that matched ShowFrom.
+	showFromLocs := make(map[uint64]bool)
+
+	// Apply to locations.
+	for _, location := range p.Location {
```
[the remark this answers is not in the document]
Comment (lvng) [reply profile/filter.go]: done.

File: profile/filter_test.go [hunk 11]
```diff
@@ -392,6 +414,113 @@ func TestFilter(t *testing.T) {
 	}
 }
 
+func TestShowFrom(t *testing.T) {
+	for _, tc := range []struct {
+		// name is the name of the test case.
```
[the remark this answers is not in the document]
Comment (lvng) [reply profile/filter_test.go]: done.

File: profile/filter_test.go [hunk 12]
```diff
@@ -392,6 +414,113 @@ func TestFilter(t *testing.T) {
...
+			showFrom:  regexp.MustCompile("map1"),
+			wantMatch: true,
+			wantSampleFuncs: []string{
+				": 2",
+			},
+		},
+	} {
+		t.Run(tc.name, func(t *testing.T) {
+			p := tc.profile.Copy()
+			gotMatch := p.ShowFrom(tc.showFrom)
```
[the remark this answers is not in the document]
Comment (lvng) [reply profile/filter_test.go]: done.

File: profile/filter_test.go [hunk 13]
```diff
@@ -392,6 +414,113 @@ func TestFilter(t *testing.T) {
...
+
+			if gotMatch != tc.wantMatch {
+				t.Errorf("match got %+v want %+v", gotMatch, tc.wantMatch)
```
[the remark this answers is not in the document]
Comment (lvng) [reply profile/filter_test.go]: done.

File: profile/filter_test.go [hunk 14]
```diff
@@ -392,6 +414,113 @@ func TestFilter(t *testing.T) {
 	}
 }
 
+func TestShowFrom(t *testing.T) {
+	for _, tc := range []struct {
```
[the remark this answers is not in the document]
Comment (lvng) [reply profile/filter_test.go]: added two test cases. one for matching multiple frames without inlines, and one with inlines.

**Comment by lvng** [comment]
PTAL.

File: internal/driver/driver.go [hunk 15]
```diff
@@ -150,11 +150,11 @@ func generateReport(p *profile.Profile, cmd []string, vars variables, o *plugin.
 }
 
 func applyCommandOverrides(cmd []string, v variables) variables {
-	trim, focus, tagfocus, hide := v["trim"].boolValue(), true, true, true
+	trim, tagfocus, filter := v["trim"].boolValue(), true, true
```
Comment (aalexand) [inline internal/driver/driver.go]: Maybe rename "tagfocus" to "tagfilter", I think it will be clearer, too.

File: profile/filter.go [hunk 16]
```diff
@@ -74,6 +74,71 @@ func (p *Profile) FilterSamplesByName(focus, ignore, hide, show *regexp.Regexp)
...
+		}
+	}
+	p.Sample = s
+	return matched
+}
+
+// filterShowFromLocation tests a showFrom regex against a location, removes
+// lines after the last match and returns whether a match was found. If the
+// mapping is matched, then all lines are kept.
+func filterShowFromLocation(loc *Location, showFrom *regexp.Regexp) (matched bool) {
```
Comment (aalexand) [inline profile/filter.go]: I think named return parameter doesn't give here much. I'd just return a bool and then do "return true" and return false" as needed.

File: profile/filter.go [hunk 17]
```diff
@@ -74,6 +74,71 @@ func (p *Profile) FilterSamplesByName(focus, ignore, hide, show *regexp.Regexp)
...
+	if m := loc.Mapping; m != nil && showFrom.MatchString(m.File) {
+		return true
+	}
+	if i := loc.lastMatchedLineIndex(showFrom); i >= 0 {
+		matched = true
+		loc.Line = loc.Line[0 : i+1]
```
Comment (aalexand) [inline profile/filter.go]: Nit: specifying zero is not required: `loc.Line[:i+1]`

File: internal/driver/driver.go (hunk 15 above)
Comment (lvng) [reply internal/driver/driver.go]: done.

File: profile/filter.go (hunk 16 above)
Comment (lvng) [reply profile/filter.go]: done.

File: profile/filter.go (hunk 17 above)
Comment (lvng) [reply profile/filter.go]: done.

File: doc/README.md [hunk 18]
```diff
@@ -94,6 +94,8 @@ Some common pprof options are:
   *regex*.
 * **-ignore= _regex_:** Do not include samples that include a report entry matching
   *regex*.
+* **-show\_from= _regex_:** Do not show entries above the first frame that
```
Comment (aalexand) [inline doc/README.md]: Maybe say "above the first one" as it doesn't look like the text here uses term "frame".

Comment (lvng) [reply doc/README.md]: done.

**Pull request merged**